runtime/compiler/env/VMJ9.cpp

```diff
@@ -2309,18 +2309,40 @@ TR_MarkHotField::markHotField(J9Class * clazz, bool rootClass)
       J9ROMClass* romClass = TR::Compiler->cls.romClassOf((TR_OpaqueClassBlock*)clazz);
       J9UTF8* name = J9ROMCLASS_CLASSNAME(romClass);
       printf("%*smarked field as hot in class %.*s\n", depth(), " ", J9UTF8_LENGTH(name), J9UTF8_DATA(name));
-      }
-
-   *(UDATA *)((char *)clazz + offsetOfHotFields()) = noncoldWord | _bitValue;
-
-   return true;
+      }  
+
+      if (jitConfig->javaVM->memoryManagerFunctions->j9gc_dynamic_breadth_first_scan_ordering_enabled(jitConfig->javaVM)) {
+          /* Setting hot fields for dynamic breadth first scan ordering */
+         if(clazz->totalInstanceSize <=512) {
+            if ((_slotIndex !=  clazz->hotField2.hotFieldOffset) &&
+               ((clazz->hotField1.hotFieldOffset == UDATA_MAX && _comp->getMethodHotness() >=warm) ||
+               (_comp->getMethodHotness() > clazz->hotField1.hotFieldMethodHotness && frequency > ((clazz->hotField1.hotFieldThreshold/10)*9)) ||
+               (_comp->getMethodHotness() == clazz->hotField1.hotFieldMethodHotness && frequency > ((clazz->hotField1.hotFieldThreshold/9)*10)))    
+            ) {
+               clazz->hotField1.hotFieldOffset = _slotIndex;
+               clazz->hotField1.hotFieldMethodHotness = _comp->getMethodHotness(); 
+               clazz->hotField1.hotFieldThreshold = frequency;
+            } else if ((_slotIndex !=  clazz->hotField1.hotFieldOffset) &&
+               ((clazz->hotField2.hotFieldOffset == UDATA_MAX  && _comp->getMethodHotness() >= hot) ||
+               (_comp->getMethodHotness() > clazz->hotField2.hotFieldMethodHotness && frequency > ((clazz->hotField2.hotFieldThreshold/10)*9)) ||
+               (_comp->getMethodHotness() == clazz->hotField2.hotFieldMethodHotness && frequency > ((clazz->hotField2.hotFieldThreshold/9)*10)))
+               ) {
+               clazz->hotField2.hotFieldOffset = _slotIndex;
+               clazz->hotField2.hotFieldMethodHotness = _comp->getMethodHotness(); 
+               clazz->hotField2.hotFieldThreshold = frequency;
+            }
+         }
```

Comment:
I roughly understand the logic here, the 3 criteria: for initial hotfild population, and override for same hotness but sligtly higher threshold (to probably avoid frequent overrides) and override with even higher hotness level that is believed to be more reliable.... I don't have anything better to suggest right now, but this is certainly subject to further investigation.

The only thing that I don't like is asymmetry for initial hotness. Why would it be different for field1 and field2?

Reply:
As a result of our conversation I will leave this as it is for now until it is reviewed from someone on the JIT side.